[roadmap] rework item on performance (#95)
coq/ceps: text/rocq-roadmap.md

This roadmap outlines our vision for Rocq's future, focusing on four key pillars:

+ **Open and Accessible**: The Rocq Prover provides user-friendly installation, comprehensive educational resources, and a thriving community to ensure everyone can benefit from its power.
+ **Trustworthy**: The Rocq Prover prioritizes self-verification, ensuring the highest level of trust in your formal proofs.

> + **Performance**: We are committed to optimizing the system for our flagship projects in both industry and academia.

Reviewer: Repetition of "We are committed to" on two subsequent lines (this one and the next that you did not change).

Furthermore, it should be "Performant" rather than "Performance" for consistency with the surrounding bullet points.

What about:

```suggestion
+ **Performant**: We optimize the system for our flagship projects in both industry and academia.
```

### 4. Enhanced Usability

> * **Performance**: optimize the system for flagship projects in both industry and academia.

Reviewer 2: "optimize" isn't quite right--we want to improve performance where it makes a difference.  There's usually a tradeoff between performance and effort/maintainability.  The real optimization is that we should optimize to deliver the most value to users with the least effort.

Merely saying "improve performance" doesn't tell the reader much (yawn)--the same words could be used for almost any software project.  It needs more (brief) Coq-specific details to be interesting to read.  One example: something like making compilation of libraries significantly faster through increased parallelism, e.g. for the benefit of users such as Bedrock.  Extra points for giving a target, e.g. delivering a typical 10x speedup for large developments.

FWIW, "Usability" is primarily about making users more efficient in performing necessary tasks, e.g. adding features to refactor proof scripts would make users more efficient/productive.

Reviewer 2: Also "flagship" is pretty vague.  I think we want to better support industrial-scale verification.  I don't know what the performance issues are for academia, whether they are much different from industry.  For academia, are there special performance needs for cutting edge research in type theory or mathematics?

Author: I'll try to give examples:

Formalizing abstract mathematics, eg category theory, stresses some components of the system, like the universes-constraint checker. I don't think this math is relevant to the industry, at least not immediately. We have ongoing work to make this better/faster. UniMath will take advantage of this.

Verifying a compiler or an imperative program via Iris stresses other components. These use cases are relevant to the software industry.

To me flagship projects are projects "you should have heard of in your area of interest". CompCert, Iris if you are a PL person, MathComp, UniMath if you do math... 

In short I just wanted to say we focus the optimization efforts towards the flagship projects and that these encompasses both industrial verification projects and more academic ones like pure math. 

About usability, I know it is a bit on the edge, but for large verification efforts performance is key to usability (in the sense of being actually usable, not only pleasant to use).

Reviewer 2: ```suggestion
* **Performance**: address performance limitations that impact users and in support
  of cutting-edge work such as:
  - large scale industrial software verification (faster build times)
  - formalization of category theory in UniComp (universe constraint checking)
  - MathComp [how would you describe this?]
  - verification of compilers and imperative programs (relying on the Iris package)
```

Maybe something like this?

Author: Thanks, my counterproposal is:

**Performance**: address performance limitations that impact our users, especially in support
  of industrial software verification and large mathematical libraries.

Let's see what others think as well

Reviewer 3: I find both proposals to be oddly specific. So, here is my take on it: "address performance limitations that impacts projects that are widely used in industry and academia".

Author: Maybe we could turn it into a more positive version, e.g. do not talk about limitation explicitly, but rather improve the perf so that things can go large(e) scale. But I'm not able to put that into (decent) words.

Reviewer 2: > I find both proposals to be oddly specific.

The roadmap would much stronger with more detail.  We can be clear that these are things we would like to address in the near future but it is not guaranteed.

At some point, perhaps in a different document, we should identify which potential improvements to Rocq will provide the most value to users, their cost and rank them (without regard to whether anyone is already committed to implement them).

Reviewer 2: > Maybe we could turn it into a more positive version, e.g. do not talk about limitation explicitly, but rather improve the perf so that things can go large(e) scale.

I suppose you could say "unmet performance needs of ...", but that's more of a euphemism, i.e., more vague and easier to misinterpret.  Every software package has performance limitations.  It's just a fact.  Saying so is not disparaging the software package.

> * **Performance**: address performance limitations that impact projects that are widely used in industry and academia.

Reviewer: For consistency with the rest of the items, there should be target audiences. Given the content of the line, this should be:
```suggestion
* **Performance**: address performance limitations that impact projects that are widely used in industry and academia (C, D).
```
